File: scripts/organize_stems.py
import json
import shutil
import wave
from datetime import datetime
from pathlib import Path
from rich_utils import print_warning, print_success, console


def organize_stems(artist, title, source_dir, stems_dir, overwrite=False):
    """Organize and validate stems with metadata generation."""
    source = Path(source_dir)
    stems_path = Path(stems_dir)

    if not source.exists():
        raise FileNotFoundError(f"Source directory not found: {source_dir}")

    stems_path.mkdir(parents=True, exist_ok=True)

    # Expected stem names
    expected_stems = ["Vocals", "Drums", "Bass", "Harmony", "Lead", "FX"]

    stems_data = {
        "track": title,
        "artist": artist,
        "export_date": datetime.now().isoformat() + "Z",
        "stems": [],
    }

    # Find and organize stem files
    for stem_name in expected_stems:
        # Look for files containing stem name
        pattern = f"*{stem_name}*"
        matching_files = list(source.glob(pattern))

        if matching_files:
            if len(matching_files) > 1:
                print_warning(f"Multiple files match '{stem_name}':")
                for f in matching_files:
                    console.print(f"     [dim]-[/dim] {f.name}")
                console.print(f"   [dim]Using:[/dim] {matching_files[0].name}")

            file = matching_files[0]
            new_name = f"{artist} - {title} - {stem_name}.wav"
            dest_file = stems_path / new_name

            # Check for existing file
            if dest_file.exists() and not overwrite:
                raise FileExistsError(
                    f"File already exists: {dest_file}\n"
                    f"  To overwrite, set 'overwrite_existing: true' in release.json"
                )

            # Copy file
            shutil.copy2(file, dest_file)

            # Get file info
            try:
                with wave.open(str(dest_file), "rb") as wav_file:
                    frames = wav_file.getnframes()
                    sample_rate = wav_file.getframerate()
                    duration = frames / float(sample_rate)
                    duration_str = f"{int(duration // 60)}:{int(duration % 60):02d}"
            except Exception as e:
                print_warning(f"Could not read duration for {new_name}: {e}")
                duration_str = "N/A"

            stems_data["stems"].append(
                {"name": stem_name, "file": new_name, "duration": duration_str}
            )
            print_success(f"Organized: {new_name}")

    # Save metadata
    metadata_file = stems_path / f"{artist} - {title} - Stems_Metadata.json"
    with open(metadata_file, "w") as f:
        json.dump(stems_data, f, indent=2)

    print_success(f"Generated metadata: {metadata_file}")
    return stems_data


if __name__ == "__main__":
    # Usage
    organize_stems(
        "YourArtistName", "Deep Dive", "./exports/stems", "./Releases/DeepDive/Stems"
    )

File: scripts/organize_stems.js
const fs = require('fs');
const path = require('path');

function organizeStems(artist, title, sourceDir, stemsDir, overwrite = false) {
  if (!fs.existsSync(sourceDir)) {
    throw new Error(`Source directory not found: ${sourceDir}`);
  }
  
  if (!fs.existsSync(stemsDir)) {
    fs.mkdirSync(stemsDir, { recursive: true });
  }
  
  const expectedStems = ["Vocals", "Drums", "Bass", "Harmony", "Lead", "FX"];
  const stemsData = {
    track: title,
    artist: artist,
    export_date: new Date().toISOString(),
    stems: []
  };
  
  const files = fs.readdirSync(sourceDir);
  
  expectedStems.forEach(stemName => {
    const matchingFiles = files.filter(f => 
      f.toLowerCase().includes(stemName.toLowerCase()) && 
      (f.endsWith('.wav') || f.endsWith('.mp3'))
    );
    
    if (matchingFiles.length > 0) {
      if (matchingFiles.length > 1) {
        console.log(`⚠️  Multiple files match '${stemName}':`);
        matchingFiles.forEach(f => console.log(`     - ${f}`));
        console.log(`   Using: ${matchingFiles[0]}`);
      }
      
      const matchingFile = matchingFiles[0];
      const newName = `${artist} - ${title} - ${stemName}.wav`;
      const sourcePath = path.join(sourceDir, matchingFile);
      const destPath = path.join(stemsDir, newName);
      
      // Check for existing file
      if (fs.existsSync(destPath) && !overwrite) {
        throw new Error(
          `File already exists: ${destPath}\n` +
          `  To overwrite, set 'overwrite_existing: true' in config.json`
        );
      }
      
      fs.copyFileSync(sourcePath, destPath);
      
      stemsData.stems.push({
        name: stemName,
        file: newName,
        duration: "N/A"
      });
      
      console.log(`✓ Organized: ${newName}`);
    }
  });
  
  // Save metadata
  const metadataFile = path.join(stemsDir, `${artist} - ${title} - Stems_Metadata.json`);
  fs.writeFileSync(metadataFile, JSON.stringify(stemsData, null, 2));
  console.log(`✓ Generated metadata: ${metadataFile}`);
  
  return stemsData;
}

// Usage
if (require.main === module) {
  organizeStems('YourArtistName', 'Deep Dive', './exports/stems', './Releases/DeepDive/Stems');
}

module.exports = { organizeStems };

